plugins/module_utils/common.py

```diff
@@ -98,112 +99,83 @@
-    'proxy': 'proxy',
-    'persist_config': 'persist_config',
-}
+def get_api_client(module=None, **auth_params):
+    auth_params = auth_params or {}
+    auth = {}
+
+    import hashlib
+    m = hashlib.sha256()
+    for i in AUTH_ARG_MAP:
+        m.update(str(auth_params.get(i)).encode())
+    digest = m.hexdigest()
+    if digest in get_api_client._pool:
+        client = get_api_client._pool[digest]
+        return client
+
+    def _raise_or_fail(exc, msg):
+        if module:
+            module.fail_json(msg % to_native(exc))
+        else:
+            raise exc
+
+    # If authorization variables aren't defined, look for them in environment variables
+    for true_name, arg_name in AUTH_ARG_MAP.items():
+        if auth_params.get(arg_name) is None:
+            env_value = os.getenv('K8S_AUTH_{0}'.format(arg_name.upper()), None) or os.getenv('K8S_AUTH_{0}'.format(true_name.upper()), None)
+            if env_value is not None:
+                if AUTH_ARG_SPEC[arg_name].get('type') == 'bool':
+                    env_value = env_value.lower() not in ['0', 'false', 'no']
+                auth[true_name] = env_value
+        else:
+            auth[true_name] = auth_params[arg_name]
+
+    def auth_set(*names):
+        return all([auth.get(name) for name in names])
+
+    if auth_set('username', 'password', 'host') or auth_set('api_key', 'host'):
+        # We have enough in the parameters to authenticate, no need to load incluster or kubeconfig
+        pass
+    elif auth_set('kubeconfig') or auth_set('context'):
+        try:
+            kubernetes.config.load_kube_config(auth.get('kubeconfig'), auth.get('context'), persist_config=auth.get('persist_config'))
+        except Exception as err:
+            _raise_or_fail(err, 'Failed to load kubeconfig due to %s')
+
+    else:
+        # First try to do incluster config, then kubeconfig
+        try:
+            kubernetes.config.load_incluster_config()
+        except kubernetes.config.ConfigException:
+            try:
+                kubernetes.config.load_kube_config(auth.get('kubeconfig'), auth.get('context'), persist_config=auth.get('persist_config'))
+            except Exception as err:
+                _raise_or_fail(err, 'Failed to load kubeconfig due to %s')
+
+    # Override any values in the default configuration with Ansible parameters
+    # As of kubernetes-client v12.0.0, get_default_copy() is required here
+    try:
+        configuration = kubernetes.client.Configuration().get_default_copy()
+    except AttributeError:
+        configuration = kubernetes.client.Configuration()
+
+    for key, value in iteritems(auth):
+        if key in AUTH_ARG_MAP.keys() and value is not None:
+            if key == 'api_key':
+                setattr(configuration, key, {'authorization': "Bearer {0}".format(value)})
+            else:
+                setattr(configuration, key, value)
+
+    try:
+        return DynamicClient(kubernetes.client.ApiClient(configuration))
```

Comment:
I think we don't want to return here and should instead put line 173 inside this block